File: Tests/test_stability.m
function test_stability()
% test_stability  Tests for stabilityTest (VLE).
%
% Cases:
%   1. Pure CH4 at 300 K / 1 bar → single-phase gas, expect STABLE.
%   2. CH4/C10H22 50:50 at 300 K / 5 MPa → two-phase expected, UNSTABLE.
%   3. result struct fields are populated correctly.
%   4. Numeric flags are consistent with result.overall.

thermo = ThermoModel();

% ---- Case 1: pure CH4 at 300 K / 1 bar (single-phase gas) ---------------
[comp1, ~] = addComponents({'CH4'});
mix1 = Mixture(comp1, 300, 1e5);  % 1 bar, well above bubble point

[flag1, SL1, SV1, res1] = stabilityTest(mix1, thermo);

pvt_assert(numel(flag1) == 2,          'stabilityTest: flag has 2 elements');
pvt_assert(SL1 >= 0,                   'stabilityTest: SL >= 0');
pvt_assert(SV1 >= 0,                   'stabilityTest: SV >= 0');
pvt_assert(ischar(res1.overall),       'result.overall is a string');
pvt_assert(ischar(res1.message),       'result.message is a string');
pvt_assert(ischar(res1.test1_message), 'result.test1_message is a string');
pvt_assert(ischar(res1.test2_message), 'result.test2_message is a string');
pvt_assert(strcmp(res1.overall, 'stable'), ...
    'pure CH4 300K/1bar: expected stable');

% Flags should be consistent with result
if strcmp(res1.overall, 'stable')
    pvt_assert(all(flag1 == 1), 'stable result: both flags == 1');
end

% ---- Case 2: CH4/C10H22 at 300 K / 5 MPa (two-phase region) -------------
[comp2, ~] = addComponents({'CH4', 'C10H22'});
mix2 = Mixture(comp2, 300, 5e6);
mix2.mole_fraction = [0.5 0.5];

[flag2, ~, ~, res2] = stabilityTest(mix2, thermo);

pvt_assert(numel(flag2) == 2, 'stabilityTest CH4/C10H22: flag has 2 elements');
pvt_assert(strcmp(res2.overall, 'unstable'), ...
    'CH4/C10H22 300K/5MPa: expected unstable');
pvt_assert(any(flag2 == 2), ...
    'CH4/C10H22 300K/5MPa: at least one test finds non-trivial solution');

% ---- Case 3: FlashOptions passed as 3rd argument -------------------------
opts = FlashOptions();
opts.maxIteration = 100;
[flag3, ~, ~, res3] = stabilityTest(mix2, thermo, opts);
pvt_assert(numel(flag3) == 2, 'stabilityTest with options: flag has 2 elements');
pvt_assert(ischar(res3.overall), 'stabilityTest with options: result.overall present');

% ---- Case 4: nargout consistency ----------------------------------------
% Just verify the function doesn't error when 0 outputs captured.
% We can't easily suppress the print in a test, so capture all outputs.
[f4, ~, ~, r4] = stabilityTest(mix1, thermo);
if strcmp(r4.overall, 'stable')
    pvt_assert(all(f4 == 1), 'nargout check: stable → all flags 1');
elseif strcmp(r4.overall, 'unstable')
    pvt_assert(any(f4 == 2), 'nargout check: unstable → a flag is 2');
end

% ---- Case 5: methanol/water below bubble point → unstable ----------------
[comp5, ~] = addComponents({'CH4O', 'H2O'});
mix5 = Mixture(comp5, 322.91, 30000);  % midrange two-phase pressure
mix5.bip.EOScons = [0 -0.07; -0.07 0];
[flag5, ~, ~, res5] = stabilityTest(mix5, thermo);
pvt_assert(numel(flag5) == 2, 'MeOH/H2O stability: flag has 2 elements');
% At these conditions we expect two phases — just check it runs and
% result.overall is one of the valid values.
pvt_assert(any(strcmp(res5.overall, {'stable','unstable','inconclusive'})), ...
    'MeOH/H2O stability: result.overall is valid string');

File: tests/test_stability.py
import numpy as np

from pvt import (
    Component,
    Mixture,
    ThermoModel,
    stabilityTest,
    stabilityLLETest,
)


def test_ch4_stable():
    comps, _ = Component.from_database_array(["CH4"])
    mix = Mixture(comps, 300.0, 1e5)
    thermo = ThermoModel()

    flag, SL, SV, result = stabilityTest(mix, thermo)

    assert result["overall"] == "stable", f"Expected stable, got {result['overall']}: {result['message']}"


def test_ch4_c10h22_unstable():
    comps, _ = Component.from_database_array(["CH4", "C10H22"])
    mix = Mixture(comps, 300.0, 5e6)
    mix.mole_fraction = np.array([0.5, 0.5])
    thermo = ThermoModel()

    flag, SL, SV, result = stabilityTest(mix, thermo)

    assert result["overall"] == "unstable", f"Expected unstable, got {result['overall']}: {result['message']}"


def test_h2o_c10h22_lle():
    comps, _ = Component.from_database_array(["H2O", "C10H22"])
    mix = Mixture(comps, 300.0, 100e5)
    mix.mole_fraction = np.array([0.5, 0.5])
    thermo = ThermoModel()

    flag, SL, SV, result = stabilityLLETest(mix, thermo)

    assert result["overall"] == "unstable", f"Expected LLE unstable, got {result['overall']}: {result['message']}"
